src/components/molecules/Visualizer/Engine/Cesium/Box/utils.ts

```diff
@@ -0,0 +1,152 @@
+  keepBoxAboveGround: boolean,
+  terrainHeightEstimate: number,
+) => {
+  if (keepBoxAboveGround) {
+    const cartographic = Cartographic.fromCartesian(trs.translation, undefined, new Cartographic());
+    const boxBottomHeight = cartographic.height - trs.scale.z / 2;
+    const floorHeight = terrainHeightEstimate;
+    if (boxBottomHeight < floorHeight) {
+      cartographic.height += floorHeight - boxBottomHeight;
+      Cartographic.toCartesian(cartographic, undefined, trs.translation);
+    }
+  }
+};
+
+export const updateTrs = (
+  trs: TranslationRotationScale,
+  property: Property | undefined,
+  terrainHeightEstimation: number,
+) => {
+  const { location, height, width, length, heading, pitch, roll, keepBoxAboveGround } =
+    property?.default ?? {};
+
+  const translation = location
+    ? Cartesian3.fromDegrees(location.lng, location.lat, location.height ?? 0)
+    : undefined;
+  if (translation) {
+    Cartesian3.clone(translation, trs.translation);
+  }
+  if (keepBoxAboveGround) {
+    translationWithClamping(trs, keepBoxAboveGround, terrainHeightEstimation);
+  }
+
+  const rotation =
+    heading && pitch && roll
+      ? Matrix3.fromHeadingPitchRoll(new HeadingPitchRoll(heading, pitch, roll))
+      : Matrix3.getRotation(
+          Matrix4.getMatrix3(Transforms.eastNorthUpToFixedFrame(trs.translation), new Matrix3()),
+          new Matrix3(),
+        );
+  Quaternion.clone(Quaternion.fromRotationMatrix(rotation), trs.rotation);
+
+  Cartesian3.clone(new Cartesian3(width || 100, length || 100, height || 100), trs.scale);
+
+  return trs;
+};
+
+function screenProjectVector(
+  scene: Scene,
+  position: Cartesian3,
+  direction: Cartesian3,
+  length: number,
+  result: Cartesian2,
+): Cartesian2 {
+  const ray = new Ray();
+  ray.origin = position;
+  ray.direction = direction;
+  const nearPoint2d = scene.cartesianToCanvasCoordinates(Ray.getPoint(ray, 0), new Cartesian2());
+
+  const farPoint2d = scene.cartesianToCanvasCoordinates(
+    Ray.getPoint(ray, length),
+    new Cartesian2(),
+  );
+  const screenVector2d = Cartesian2.subtract(farPoint2d, nearPoint2d, result);
+  return screenVector2d;
+}
+
+const dotProductMousePosition = (
+  scene: Scene,
+  mouseMove: {
+    startPosition: Cartesian2;
+    endPosition: Cartesian2;
+  },
+  position: Cartesian3,
+  direction: Cartesian3,
+) => {
+  const mouseVector2d = Cartesian2.subtract(
+    mouseMove.endPosition,
+    mouseMove.startPosition,
+    new Cartesian2(),
+  );
+
+  // Project the vector of unit length to the screen
+  const screenVector2d = screenProjectVector(scene, position, direction, 1, new Cartesian2());
+  const screenNormal2d = Cartesian2.normalize(screenVector2d, new Cartesian2());
+
+  const pixelsPerStep = Cartesian2.magnitude(screenVector2d);
+  const moveAmountPixels = Cartesian2.dot(mouseVector2d, screenNormal2d);
+
+  return {
+    pixelsPerStep,
+    moveAmountPixels,
+  };
+};
+
+export const computeMoveAmount = (
```

Comment:
It'll be kinda long, but should this be `computeMouseMoveAmount`? Or `computeMouseMove`?

Reply:
Exactly, I fixed this in [refactor: rename computeMoveAmount](https://github.com/reearth/reearth-web/pull/357/commits/f89ea5836794aa4d5a6a570da3ab4ae080462421)